library/core/lifecycle-events/src/main/java/org/quiltmc/qsl/lifecycle/api/client/event/ClientLifecycleEvents.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.lifecycle.api.client.event;
+
+import org.quiltmc.qsl.base.api.event.ArrayEvent;
+
+import net.minecraft.client.MinecraftClient;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+
+/**
+ * Events indicating the lifecycle of a Minecraft client.
+ *
+ * <p>The lifecycle of a Minecraft client, starts when the client is ready. The client will tick the client and then the
+ * integrated server if in single player.
+ */
+@Environment(EnvType.CLIENT)
+public final class ClientLifecycleEvents {
+	// There is no STARTING event because there is no way to allow mods to register callbacks that early without possibly
+	// initializing the game's registries improperly in preLaunch.
+
+	/**
+	 * An event indicating that a Minecraft client is ready to tick and render.
+	 *
+	 * <p>It should be noted this event is executed while the splash screen is visible, not the main menu.
+	 */
+	public static final ArrayEvent<Ready> READY = ArrayEvent.create(Ready.class, callbacks -> client -> {
+		for (var callback : callbacks) {
+			callback.readyClient(client);
+		}
+	});
+
+	/**
+	 * An event indicating that a Minecraft client has finished it's last tick and will shut down.
+	 *
+	 * <p>If the client is connected to a server, the client will disconnect from the server. Then if the client has a
+	 * running an integrated server, the integrated server will be shutdown. Finally all client facilities are torn down.
```

Comment:
does this happen before or after? also the documentation sorta sounds like disconnecting would fire this event.

Reply:
I think it happens before iirc

Reply:
Fired before, disconnection occurs as a result of exiting the client.

Reply:
> Fired before, disconnection occurs as a result of exiting the client.

ah, might be worth mentioning it in the docs then, as else it's confusing